crates/ruma-server-util/src/authorization.rs

```diff
@@ -0,0 +1,284 @@
+    /// the request.
+    pub key: OwnedServerSigningKeyId,
+    /// The signature of the JSON.
+    pub sig: String,
+}
+
+impl XMatrix {
+    /// Construct a new X-Matrix Authorization header.
+    pub fn new(
+        origin: OwnedServerName,
+        destination: Option<OwnedServerName>,
+        key: OwnedServerSigningKeyId,
+        sig: String,
+    ) -> Self {
+        Self { origin, destination, key, sig }
+    }
+}
+
+fn parse_token<'a>(tokens: &mut impl Tokens<Item = &'a u8>) -> Option<Vec<u8>> {
+    tokens.optional(|t| {
+        let token: Vec<u8> = t.tokens_while(|c| is_tchar(**c)).copied().collect();
+        if !token.is_empty() {
+            Some(token)
+        } else {
+            debug!("Returning early because of empty token at {}", t.location().offset());
+            None
+        }
+    })
+}
+
+fn parse_token_with_colons<'a>(tokens: &mut impl Tokens<Item = &'a u8>) -> Option<Vec<u8>> {
+    tokens.optional(|t| {
+        let token: Vec<u8> = t.tokens_while(|c| is_tchar(**c) || **c == b':').copied().collect();
+        if !token.is_empty() {
+            Some(token)
+        } else {
+            debug!("Returning early because of empty token at {}", t.location().offset());
+            None
+        }
+    })
+}
+
+fn parse_quoted<'a>(tokens: &mut impl Tokens<Item = &'a u8>) -> Option<Vec<u8>> {
+    tokens.optional(|t| {
+        if !(t.token(&b'"')) {
+            return None;
+        }
+        let mut buffer = Vec::new();
+        loop {
+            match t.next()? {
+                // quoted pair
+                b'\\' => {
+                    let escaped = t.next().filter(|c| {
+                        if is_quoted_pair(**c) {
+                            true
+                        } else {
+                            debug!(
+                                "Encountered an illegal character {} at location {}",
+                                **c as char,
+                                t.location().offset()
+                            );
+                            false
+                        }
+                    })?;
+                    buffer.push(*escaped)
+                }
+                // end of quote
+                b'"' => break,
+                // regular character
+                c if is_qdtext(*c) => buffer.push(*c),
+                // Invalid character
+                c => {
+                    debug!(
+                        "Encountered an illegal character {} at location {}",
+                        *c as char,
+                        t.location().offset()
+                    );
+                    return None;
+                }
+            }
+        }
+        Some(buffer)
+    })
+}
+
+fn parse_xmatrix_field<'a>(tokens: &mut impl Tokens<Item = &'a u8>) -> Option<(String, Vec<u8>)> {
+    tokens.optional(|t| {
+        let name = parse_token(t).and_then(|name| {
+            let name = dbg!(std::str::from_utf8(&name).ok()?.to_ascii_lowercase());
```

Comment:
I'm curious, why does this use `dbg!`?

Reply:
ah, that's a leftover from writing this, when I had a problem parsing. Thanks for catching this, will remove it.

Reply:
TIL that `dbg!` prints _and_ lets you assign the value, hence why I wasn't sure whether it might be there on purpose.

Reply:
Yeah, `dbg!` is amazing, you can wrap it around basically anything without changing much of your code and capture what an expression is evaluated to.